### DISCUSSION TITLE
NextJS buttons are not actionable right away (potentially related to hydration)

### DISCUSSION THREAD

**Discussion by cachho** [cypress-io/cypress#30631] [Questions and Help]
I have a pretty big app, meaning a big layout, many providers etc. (that's probably why I couldn't minimally reproduce this).

The problem: I want to test the action of a button in a client component.

**The problem: when buttons are first rendered in, clicking them does nothing; cypress clicks them right when they appear. So my tests fail.**

This behavior cannot be reproduced manually, because I'm not able to click the button instantly.

The obvious workaround is to just have cypress wait arbitrarily, this works, but it's not recommended for reasons that have been discussed plenty of times.

The other workaround I found is to disable the button until it is hydrated on the client via `useEffect`

```ts
const [ready, setReady] = React.useState(false)

useEffect(() => {
  setReady(true)
, [])

return (<button onClick={() => {}} disabled={!isReady}>Button</button>)
```

I don't want to do this because I think this has SEO disadvantages, as the initial render now shows disabled button. I also have to write my code around the tests, which should not be the case.

This issue does not only apply to buttons. I'm also having this issue with input fields that cypress can't type into, giving me an error that they are disabled, when they certainly aren't disabled.

Am I the only one with this problem? What would be the best way to solve it? Can someone point me in the right direction?

**Comment by portyu9** [comment]
Hi @cachho 

Since this is Component Testing, I wouldn't assume this is SSR hydration without a reproduction showing hydration errors.

What the behavior does show is that the DOM becomes actionable before some application/provider initialization is complete. Cypress can wait for DOM actionability, but it cannot know when an application-specific initialization step has finished.

I would synchronize on a deterministic signal that proves the component is ready—for example, the provider's initial request completing, a loading state disappearing, or initialized content appearing—rather than using a fixed cy.wait().

If there is genuinely no observable state distinguishing “rendered” from “ready to interact,” that is an application race rather than something Cypress can generically detect. In that case, exposing a real ready/loading state is the robust fix for both tests and users.

**Reply by portyu9** [reply]
@cachho Please mark this answer as accepted if it helped, thank you.

**Reply by cachho** [reply]
@portyu9 I switched away from Cypress. Judging from the upvotes I guess that other people are having the same issue, but sure I will mark it as accepted.
